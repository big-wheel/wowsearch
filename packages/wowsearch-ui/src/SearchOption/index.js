/**
 * @file index
 * @author imcuttle <moyuyc95@gmail.com>
 * @date 2019/9/26
 *
 */
import React from 'react'
import p from 'prefix-classname'

const c = p('')
const cn = p('wowsearch-search-option-')

export default function SearchOption({className, openInNew, title, crumbs = [], url, content}) {
  return (
    <a onClick={evt => {
      if (!evt.metaKey) {
        // 不走 a 默认的跳转逻辑
        evt.preventDefault()
      } else {
        evt.stopPropagation()
      }
    }} target={openInNew ? '_blank' : '_self'} href={url} className={c(cn('wrapper'), className)}>
      <div className={cn('left')} dangerouslySetInnerHTML={{__html: title}} />
      <div className={cn('right')}>
        <div className={cn('header')}>
          {crumbs.map((crumb, i) => {
            const preCrumb = i > 0 ? ' > ' : ''
            return <span key={i} className={cn('crumb')} dangerouslySetInnerHTML={{__html: preCrumb + crumb}} />
          })}
        </div>
        <div className={cn('body')} dangerouslySetInnerHTML={{__html: content}} />
      </div>
    </a>
  )
}
