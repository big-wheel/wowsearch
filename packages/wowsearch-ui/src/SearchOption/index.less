@prefix: ~'wowsearch-search-option-';

.@{prefix} {
  &wrapper {
    display: flex;
  }

  &left {
    flex: 1;
  }

  &right {
    flex: 3;
    display: flex;
    flex-direction: column;
  }
}
