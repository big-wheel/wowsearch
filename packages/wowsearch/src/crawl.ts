/**
 * @file crawl
 * @author Cuttle Cong
 * @date 2018/6/10
 * @description
 */
import { CrawlerConfig } from './types/Config'
import * as request from 'request-promise-native'
import * as select from 'xpath.js'
import { DOMParser } from 'xmldom'
import * as cheerio from 'cheerio'
import * as each from 'lodash.foreach'
import makeCheck from './makeCheckUrl'
import * as u from 'url'

const debug = require('debug')('wowsearch:crawl')

const protocolPorts = new Map<any, number>([
  ['https:', 443],
  ['http:', 80],
  [null, null]
])

export function isSameOrigin(valueUrl: string, fromUrl?: string) {
  valueUrl = valueUrl.trim()
  let { pathname, protocol, hostname, port } = u.parse(valueUrl)
  if (valueUrl.startsWith('//') || protocol == null) {
    return true
  }
  if (!fromUrl) {
    return false
  }
  const fo = u.parse(fromUrl)
  fo.port = fo.port || <any>protocolPorts.get(fo.protocol)
  port = port || <any>protocolPorts.get(protocol)
  return (
    fo.protocol === protocol && fo.hostname === hostname && fo.port === port
  )
}

export function crawl(
  text: string,
  config: CrawlerConfig,
  fromUrl = ''
): { smartCrawlingUrls; crawlTexts } {
  const $ = cheerio.load(text)
  let doc
  const {
    start_urls,
    stop_urls,
    strip_chars,
    selectors,
    selectors_exclude,
    smart_crawling,
    force_crawling_urls
  } = config
  const check = makeCheck({ start_urls, stop_urls })

  let smartCrawlingUrls = []
  if (smart_crawling) {
    $('a').map(function() {
      let href = $(this).attr('href')
      if (href && isSameOrigin(href, fromUrl)) {
        href = u.resolve(fromUrl, href)
        let o = u.parse(href)
        delete o.query
        delete o.search
        delete o.hash
        href = u.format(o)

        debug('checked href: %s, isSameOrigin', href)
        if (force_crawling_urls || check(href)) {
          fromUrl !== href &&
            smartCrawlingUrls.indexOf(href) < 0 &&
            smartCrawlingUrls.push(href)
        }
      } else {
        debug('checked href: %s, not matched', href)
      }
    })

    debug('smartCrawlingUrls: %o, from url: %s', smartCrawlingUrls, fromUrl)
  }

  // TODO
  // Do filter about start_urls and stop_urls.

  if (selectors_exclude) {
    each(selectors_exclude, (val, key) => {
      $(val).remove()
    })
  }

  let crawlTexts = []
  // { lvl0:
  //       lvl1:
  //       lvl2:
  //       lvl3:
  //       lvl4: null,
  //       lvl5: null,
  //       lvl6: null,
  //       text:

  // selectors.

  each(selectors, (value, key) => {
    let text = null
    if (typeof value === 'string') {
      value = {
        type: 'css',
        selector: value
      }
    }
    if (value.type === 'css' || !value.type) {
      text = $(value.selector).text()
      debug('css selector: %s, matched: %o.', value.selector, text)
    } else {
      if (!doc) {
        doc = new DOMParser({
          errorHandler: {
            warning: null,
            error: console.error,
            fatalError: console.error
          }
        }).parseFromString($.html())
      }

      let nodes = select(doc, value.selector)
      text = nodes.map(node => node.textContent).join('')
      debug('xpath selector: %s, matched: %o.', value.selector, text)
    }
    let strip_chars = value.strip_chars || config.strip_chars || ''
    crawlTexts[key] = text
      ? text.replace(
          new RegExp(`(^[${strip_chars}]+)|(${strip_chars}]+$)`, 'g'),
          ''
        )
      : value.hasOwnProperty('default_value')
        ? value.default_value
        : null
  })

  return { crawlTexts, smartCrawlingUrls }
}

export async function crawlByUrl(url: string, config: CrawlerConfig) {
  let html
  if (config.js_render) {
    const puppeteer = require('puppeteer')
    const browser = await puppeteer.launch({
      timeout: 100000,
      args: ['--no-sandbox', '--disable-setuid-sandbox']
    })
    const page = await browser.newPage()
    await page.goto(url, { waitUntil: 'networkidle2', timeout: 100000 })
    config.js_waitfor && (await page.waitFor(config.js_waitfor))
    html = await page.evaluate(() => {
      return document.documentElement.outerHTML
    })
    await browser.close()
  } else {
    try {
      html = await request(encodeURI(decodeURI(url)))
    } catch (e) {
      console.error(`URL: ${url}`, String(e))
      return {}
    }
  }

  let { crawlTexts, smartCrawlingUrls } = await crawl(html, config, url)
  debug('url: %s, crawlTexts: %o', url, crawlTexts)
  return { crawlTexts, smartCrawlingUrls }
}
