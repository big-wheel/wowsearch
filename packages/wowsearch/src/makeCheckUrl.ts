import { isRule, Rule } from 'wowsearch-parse/dist/match'
import match from 'wowsearch-parse/dist/match'

/**
 * @file makeCheckUrl
 * @author Cuttle Cong
 * @date 2018/6/10
 * @description
 */

export function makeIsInStartUrls(start_urls) {
  return function isInStartUrls(url) {
    return start_urls.some(rule => {
      if (isRule(rule)) {
        return match(<Rule>rule, url)
      }
      return match((<any>rule).url, url)
    })
  }
}

export default function check(start_urls = [], stop_urls = []) {
  const isInStartUrls = makeIsInStartUrls(start_urls)
  const isInStopUrls = makeIsInStopUrls(stop_urls)
  return function check(url) {
    return isInStartUrls(url) && !isInStopUrls(url)
  }
}

export function makeIsInStopUrls(stop_urls) {
  return function isInStopUrls(url) {
    return stop_urls.some(rule => {
      return match(<Rule>rule, url)
    })
  }
}
