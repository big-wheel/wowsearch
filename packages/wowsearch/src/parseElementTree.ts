import * as uniq from 'lodash.uniq'
import * as visitTree from '@moyuyc/visit-tree'

import {
  Config,
  CrawlerConfig,
  Selectors,
  Selector,
  StrictSelector
} from './types/Config'
import selectVal, {
  normalizeSelector,
  selectAll,
  selectOne,
  transformVal
} from './selectVal'
import DocumentNode from './types/DocumentNode'
import LvlNode from './types/LvlNode'
import TextNode from './types/TextNode'

export function parseLvlTypeLevel(type: string = ''): number {
  type = type.trim()
  if (/^lvl([\d+])$/.test(type)) {
    return Number(RegExp.$1)
  }
  return null
}

export function isLvlType(type: string): boolean {
  return parseLvlTypeLevel(type) !== null
}

export const LVL_TYPES = [
  'lvl0',
  'lvl1',
  'lvl2',
  'lvl3',
  'lvl4',
  'lvl5',
  'lvl6'
]

function matchSelector(
  node: Element,
  selectorList: Selector[]
): { node: Element; selector: Selector; selectorIndex: number } | null {
  let selectorNode = node
  let selectorIndex = selectorList.findIndex(s => {
    s = normalizeSelector(s)
    if (s.type === 'css') {
      if (typeof node.matches === 'function' && node.matches(s.selector)) {
        selectorNode = node
        return true
      }
      selectorNode = selectOne(node, s)
      return !!selectorNode
    } else {
      selectorNode = selectOne(node, s)
      return !!selectorNode
    }
  })

  if (selectorIndex >= 0) {
    return {
      selector: selectorList[selectorIndex],
      selectorIndex,
      node: selectorNode
    }
  }

  return null
}

function getAnchor(elem: Element, selectorItem: StrictSelector) {
  const {
    anchor_selector = 'a[id]',
    anchor_attribute_name = 'id'
  } = selectorItem
  let anchor = elem.getAttribute(anchor_attribute_name)
  if (!anchor) {
    let tmpNode = selectOne(elem, anchor_selector)
    if (tmpNode) {
      return tmpNode.getAttribute(anchor_attribute_name)
    }
  }
  return anchor
}

function generateTextNode(node: Node, selector: StrictSelector) {
  return new TextNode(transformVal(node.textContent, selector))
}

// todo:
// 1. walk document by pre order
// 2. find node which matches one of selectors
function generateLvlNode(
  selectorKey,
  startElem: Element,
  selectors: Selectors,
  walk?: Function
): {lvlNode: LvlNode, endElement: Element} {
  const selectorItem = normalizeSelector(selectors[selectorKey])
  if (!selectorItem) return null
  const level = parseLvlTypeLevel(selectorKey)
  if (level === null) {
    // new TextNode(transformVal(current.textContent, selectorItem))
    return null
  }

  const selectorKeys = uniq(LVL_TYPES.concat(Object.keys(selectors)))
  const selectorEntryList = selectorKeys.map(key => ({key, selector: selectors[key]})).filter(({selector}) => selector)
  const selectorList = selectorEntryList.map(({selector}) => selector)
  const lvlNode = new LvlNode(level)

  walk && walk(startElem)
  lvlNode.value = transformVal(startElem.textContent, selectorItem)
  lvlNode.anchor = getAnchor(startElem, selectorItem)
  const children = lvlNode.children

  let callbackList = []
  let curr = startElem.nextElementSibling
  while (curr) {
    let isBreak
    let isContinue

    runUntilEq((current) => {
      let matches
      if ((matches = matchSelector(current, selectorList))) {
        const matchedSelectorKey = selectorEntryList[matches.selectorIndex] && selectorEntryList[matches.selectorIndex].key
        if (isLvlType(matchedSelectorKey)) {
          if (level < parseLvlTypeLevel(matchedSelectorKey)) {
            // callbackList.forEach(fn => fn())
            // callbackList = []

            const childObj = generateLvlNode(matchedSelectorKey, matches.node, selectors, walk)
            children.push(childObj.lvlNode)
            // matches.node.remove()
            curr = childObj.endElement
            isContinue = true
            return
            // return matches.node
          } else {
            isBreak = true
            return
            // return matches.node
          }
        } else if (matchedSelectorKey === 'text') {
          walk && walk(matches.node)
          children.push(generateTextNode(matches.node, matches.selector))
          callbackList.push(() => {
            matches.node.remove()
          })

        }

        return matches.node
      }
    }, curr)

    if (isBreak) {
      break
    }
    if (isContinue) {
      continue
    }

    // curr && curr.remove()
    curr = curr && curr.nextElementSibling
  }

  callbackList.forEach(fn => fn())

  return {
    lvlNode,
    endElement: curr
  }
}

function runUntilEq(walk, node: Element) {
  let next = walk(node)
  // node.remove && node.remove()

  while (next && next !== node) {
    next.remove()
    // node = next
    next = walk(node)
  }
}

export default function parseElementTree(
  document: Element,
  selectors: Selectors
): DocumentNode {
  const documentNode = new DocumentNode()
  const selectorKeys = uniq(LVL_TYPES.concat(Object.keys(selectors)))

  // Deals with global selector
  selectorKeys.forEach(selectorKey => {
    const selectorItem = normalizeSelector(selectors[selectorKey])
    if (!selectorItem) return

    if (selectorItem.global) {
      documentNode.global.set(
        selectorKey,
        selectVal(selectorItem, document).text
      )
    }
  })

  const selectorList = selectorKeys
    .map(key => ({
      key,
      selector: normalizeSelector(selectors[key])
    }))
    .filter(({ selector }) => selector && !selector.global)
  const selectorListPure = selectorList.map(({ selector }) => selector)

  const track = new WeakMap()
  visitTree(
    document,
    () => {},
    (childNode: Element, ctx) => {
      // Removed or walked
      if (track.has(childNode) || !childNode.parentElement) {
        return ctx.skip()
      }

      const matched = matchSelector(childNode, selectorListPure)
      if (matched) {
        if (matched.selector) {
          const matchedNode = matched.node
          if (track.has(matchedNode)) {
            return
          }

          const { selector, key } = selectorList[matched.selectorIndex]
          if (selector.type) {
            if (key === 'text') {
              documentNode.children.push(
                generateTextNode(matchedNode, selector)
              )
              // ctx.skip()
            } else if (isLvlType(key)) {
              const {lvlNode, endElement} = generateLvlNode(key, matchedNode, selectors, (elem) => {
                track.set(elem, true)
              })
              documentNode.children.push(lvlNode)
              // ctx.skip()
            }
          }
        }
        ctx.node.remove()
        return matched.node.remove()
      }
      else {
        ctx.skip()
        // childNode.remove && childNode.remove()
      }
    },
    { path: 'children' }
  )

  return documentNode
}
