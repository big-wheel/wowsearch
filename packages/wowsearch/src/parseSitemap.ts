/**
 * @file parseSitemap
 * @author Cuttle Cong
 * @date 2018/6/10
 * @description
 */
import * as request from 'request-promise-native'
import { parseString } from 'xml2js'
import * as pify from 'pify'
import * as uniq from 'array-uniq'

export function parseSitemapXML(text: string): Promise<any> {
  return pify(parseString)(text)
}

export function parseSitemapTXT(text: string): string[] {
  return text
    .split('\n')
    .map(x => x.trim())
    .filter(Boolean)
}

export default async function parseFromUrl(url: string) {
  const text = await request(url)
  if (url.toLowerCase().endsWith('.xml')) {
    let obj = await parseSitemapXML(text)
    let urls = []
    obj.urlset.url.forEach(({ loc }) => {
      if (Array.isArray(loc)) {
        urls.push(...loc)
      }
      else {
        urls.push(loc)
      }
    })
    return uniq(urls)
  }
  return parseSitemapTXT(text)
}
