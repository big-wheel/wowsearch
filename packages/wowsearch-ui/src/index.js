/**
 * @file index
 * @author imcuttle <moyuyc95@gmail.com>
 * @date 2019/9/26
 *
 */
import React from 'react'
import Select, { Option, OptGroup } from 'rc-select'
import './index.less'
import { selectPrefix } from '!get-less-vars/loader!./index.less'
import * as PropTypes from 'prop-types'
import {robust} from 'memoize-fn'
import SearchOption from './SearchOption'

SearchUI.propTypes = {
  fetcher: PropTypes.func.isRequired
}

export default function SearchUI({ fetcher }) {
  const mFetcher = React.useMemo(() => {
    return robust(fetcher, {})
  }, [fetcher])

  const [dataSource, setDataSource] = React.useState([])
  const fetchData = async value => {
    // todo error handler
    const list = await mFetcher(value)
    setDataSource(list)
  }

  const onChange = value => {

  }
  const options = dataSource.map(d => {
    return (
      <Option key={d.id}>
        <SearchOption {...d}/>
      </Option>
    )
  })

  return (
    <Select
      labelInValue
      prefixCls={selectPrefix}
      onSearch={fetchData}
      showSearch
      value={null}
      optionLabelProp="children"
      placeholder="placeholder"
      defaultActiveFirstOption
      style={{ width: 500 }}
      onChange={onChange}
      filterOption={false}
    >
      {options}
    </Select>
  )
}
