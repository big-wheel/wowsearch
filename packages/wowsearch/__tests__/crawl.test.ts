/**
 * @file main.test
 * @author Cuttle Cong
 * @date 2018/6/10
 * @description
 */
import { readFileSync } from 'fs'
import { makeFixture } from './help'
import { crawl, isSameOrigin, crawlByUrl } from '../src/crawl'

describe('crawl', function() {
  it('crawl text', async function() {
    const texts = await crawl(
      readFileSync(makeFixture('text.html')).toString(),
      {
        strip_chars: 'r',
        selectors: {
          lvl0: '#readme .package-name-redundant',
          lvl1: '#readme h1:not(.package-name-redundant)',
          lvl2: {
            selector: '//*[@id="readme"]//h2',
            type: 'xpath'
          },
          lvl3: {
            selector: '#readme h3'
          },
          lvl4: { selector: '#readme h4', default_value: 'abc' },
          text: '#readme p, #readme li, #readme code'
        },
        selectors_exclude: ['#readme .deep-link']
      }
    )

    expect(texts).toMatchSnapshot()
  })

  it('crawl text when smart_crawling', async function() {
    const texts = await crawl(
      readFileSync(makeFixture('text.html')).toString(),
      {
        smart_crawling: true,
        start_urls: ['https://npmjs.com/settings/**'],
        selectors: {
          lvl0: '#readme .package-name-redundant',
          lvl1: '#readme h1:not(.package-name-redundant)',
          lvl2: {
            selector: '//*[@id="readme"]//h2',
            type: 'xpath'
          },
          lvl3: {
            selector: '#readme h3'
          },
          lvl4: { selector: '#readme h4', default_value: 'abc' },
          text: '#readme p, #readme li, #readme code'
        },
        selectors_exclude: ['#readme .deep-link']
      },
      'https://npmjs.com/package/robots'
    )
  })

  it('crawl text when smart_crawling and force_crawling_urls', async function() {
    const texts = await crawl(
      readFileSync(makeFixture('text.html')).toString(),
      {
        smart_crawling: true,
        force_crawling_urls: true,
        start_urls: [],
        selectors: {
          lvl0: '#readme .package-name-redundant',
          lvl1: '#readme h1:not(.package-name-redundant)',
          lvl2: {
            selector: '//*[@id="readme"]//h2',
            type: 'xpath'
          },
          lvl3: {
            selector: '#readme h3'
          },
          lvl4: {selector: '#readme h4', default_value: 'abc'},
          text: '#readme p, #readme li, #readme code'
        },
        selectors_exclude: ['#readme .deep-link']
      },
      'https://npmjs.com/package/robots'
    )

    expect(texts).toMatchSnapshot()
  })

  it('crawl text with js_render', async function() {
    jest.setTimeout(10000)
    const texts = await crawlByUrl('https://www.npmjs.com/package/the-answer', {
      js_render: true,
      selectors: {
        lvl0: '#readme .package-name-redundant',
        lvl1: '#readme h1:not(.package-name-redundant)',
        lvl2: {
          selector: '//*[@id="readme"]//h2',
          type: 'xpath'
        },
        lvl3: {
          selector: '#readme h3'
        },
        lvl4: '#readme h4',
        text: '#readme p, #readme li, #readme code'
      },
      selectors_exclude: ['#readme .deep-link']
    })

    expect(texts).toMatchSnapshot()
  })

  it('should crawl nothing', function() {
    const texts = crawl(readFileSync(makeFixture('text.html')).toString(), {
      selectors: {}
    })

    expect(texts).toMatchSnapshot()
  })

  it('should isSameOrigin', function() {
    expect(isSameOrigin('/same')).toBeTruthy()
    expect(isSameOrigin('//same')).toBeTruthy()
    expect(isSameOrigin('http://same')).toBeFalsy()
  })

  it('should isSameOrigin when fromUrl', function () {
    expect(isSameOrigin('http://www.baidu.com', 'http://www.baidu.com')).toBeTruthy()
    expect(isSameOrigin('http://www.baidu.com/wdwwa', 'http://www.baidu.com/ajha')).toBeTruthy()
  });
})
