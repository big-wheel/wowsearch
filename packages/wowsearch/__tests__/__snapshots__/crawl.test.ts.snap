// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`crawl crawl text 1`] = `
Object {
  "lvl0": "obots.js",
  "lvl1": "Resources",
  "lvl2": "InstallationUsageAPILicense",
  "lvl3": null,
  "lvl4": "abc",
  "text": "obots.js — is parser for robots.txt files for node.js.It's recommended to install via npm:Here's an example of using robots.js:Default crawler user-agent is:Mozilla/5.0 (X11; Linux i686; rv:5.0) Gecko/20100101 Firefox/5.0
Here's an example of using another user-agent and more detailed callback:Here's an example of getting list of sitemaps:Here's an example of getCrawlDelay usage:An example of passing options to the HTTP request:RobotsParser — main class. This class provides a set of methods to read,
    parse and answer questions about a single robots.txt file.setUrl(url, read) — sets the URL referring to a robots.txt file.
      by default, invokes read() method.
      If read is a function, it is called once the remote file is downloaded and parsed, and it
      takes in two arguments: the first is the parser itself, and the second is a boolean
      which is True if the the remote file was successfully parsed.read(after_parse) — reads the robots.txt URL and feeds it to the parserparse(lines) — parse the input lines from a robots.txt filecanFetch(userAgent, url, callback) — using the parsed robots.txt decide if
      userAgent can fetch url. Callback function:
      function callback(access, url, reason) { ... }
      where:
      
        access — can this url be fetched. true/false.
        url — target url
        reason — reason for access. Object:
          
            type — valid values: 'statusCode', 'entry', 'defaultEntry', 'noRule'
            entry — an instance of lib/Entry.js:. Only for types: 'entry', 'defaultEntry'
            statusCode — http response status code for url. Only for type 'statusCode'
          
        
      
    function callback(access, url, reason) { ... }access — can this url be fetched. true/false.url — target urlreason — reason for access. Object:
          
            type — valid values: 'statusCode', 'entry', 'defaultEntry', 'noRule'
            entry — an instance of lib/Entry.js:. Only for types: 'entry', 'defaultEntry'
            statusCode — http response status code for url. Only for type 'statusCode'
          
        accesstype — valid values: 'statusCode', 'entry', 'defaultEntry', 'noRule'entry — an instance of lib/Entry.js:. Only for types: 'entry', 'defaultEntry'lib/Entry.js:statusCode — http response status code for url. Only for type 'statusCode'canFetchSync(userAgent, url) — using the parsed robots.txt decide if
      userAgent can fetch url. Return true/false.getCrawlDelay(userAgent) — returns Crawl-delay for the certain userAgentgetSitemaps(sitemaps) — gets Sitemaps from parsed robots.txtgetDisallowedPaths(userAgent) — gets paths explictly disallowed for the user agent specified AND *See LICENSE
    file.Robots.txt Specifications by GoogleRobots.txt parser for pythonA Standard for Robot Exclusion",
}
`;

exports[`crawl crawl text when smart_crawling 1`] = `
Object {
  "crawlTexts": Object {
    "lvl0": "robots.js",
    "lvl1": "Resources",
    "lvl2": "InstallationUsageAPILicense",
    "lvl3": null,
    "lvl4": "abc",
    "text": "robots.js — is parser for robots.txt files for node.js.It's recommended to install via npm:Here's an example of using robots.js:Default crawler user-agent is:Mozilla/5.0 (X11; Linux i686; rv:5.0) Gecko/20100101 Firefox/5.0
Here's an example of using another user-agent and more detailed callback:Here's an example of getting list of sitemaps:Here's an example of getCrawlDelay usage:An example of passing options to the HTTP request:RobotsParser — main class. This class provides a set of methods to read,
    parse and answer questions about a single robots.txt file.setUrl(url, read) — sets the URL referring to a robots.txt file.
      by default, invokes read() method.
      If read is a function, it is called once the remote file is downloaded and parsed, and it
      takes in two arguments: the first is the parser itself, and the second is a boolean
      which is True if the the remote file was successfully parsed.read(after_parse) — reads the robots.txt URL and feeds it to the parserparse(lines) — parse the input lines from a robots.txt filecanFetch(userAgent, url, callback) — using the parsed robots.txt decide if
      userAgent can fetch url. Callback function:
      function callback(access, url, reason) { ... }
      where:
      
        access — can this url be fetched. true/false.
        url — target url
        reason — reason for access. Object:
          
            type — valid values: 'statusCode', 'entry', 'defaultEntry', 'noRule'
            entry — an instance of lib/Entry.js:. Only for types: 'entry', 'defaultEntry'
            statusCode — http response status code for url. Only for type 'statusCode'
          
        
      
    function callback(access, url, reason) { ... }access — can this url be fetched. true/false.url — target urlreason — reason for access. Object:
          
            type — valid values: 'statusCode', 'entry', 'defaultEntry', 'noRule'
            entry — an instance of lib/Entry.js:. Only for types: 'entry', 'defaultEntry'
            statusCode — http response status code for url. Only for type 'statusCode'
          
        accesstype — valid values: 'statusCode', 'entry', 'defaultEntry', 'noRule'entry — an instance of lib/Entry.js:. Only for types: 'entry', 'defaultEntry'lib/Entry.js:statusCode — http response status code for url. Only for type 'statusCode'canFetchSync(userAgent, url) — using the parsed robots.txt decide if
      userAgent can fetch url. Return true/false.getCrawlDelay(userAgent) — returns Crawl-delay for the certain userAgentgetSitemaps(sitemaps) — gets Sitemaps from parsed robots.txtgetDisallowedPaths(userAgent) — gets paths explictly disallowed for the user agent specified AND *See LICENSE
    file.Robots.txt Specifications by GoogleRobots.txt parser for pythonA Standard for Robot Exclusion",
  },
  "smartCrawlingUrls": Array [
    "https://npmjs.com/settings/moyuyc/profile",
    "https://npmjs.com/settings/moyuyc/packages",
    "https://npmjs.com/settings/moyuyc/billing",
    "https://npmjs.com/settings/moyuyc/tokens",
  ],
}
`;

exports[`crawl crawl text when smart_crawling and force_crawling_urls 1`] = `
Object {
  "crawlTexts": Object {
    "lvl0": "robots.js",
    "lvl1": "Resources",
    "lvl2": "InstallationUsageAPILicense",
    "lvl3": null,
    "lvl4": "abc",
    "text": "robots.js — is parser for robots.txt files for node.js.It's recommended to install via npm:Here's an example of using robots.js:Default crawler user-agent is:Mozilla/5.0 (X11; Linux i686; rv:5.0) Gecko/20100101 Firefox/5.0
Here's an example of using another user-agent and more detailed callback:Here's an example of getting list of sitemaps:Here's an example of getCrawlDelay usage:An example of passing options to the HTTP request:RobotsParser — main class. This class provides a set of methods to read,
    parse and answer questions about a single robots.txt file.setUrl(url, read) — sets the URL referring to a robots.txt file.
      by default, invokes read() method.
      If read is a function, it is called once the remote file is downloaded and parsed, and it
      takes in two arguments: the first is the parser itself, and the second is a boolean
      which is True if the the remote file was successfully parsed.read(after_parse) — reads the robots.txt URL and feeds it to the parserparse(lines) — parse the input lines from a robots.txt filecanFetch(userAgent, url, callback) — using the parsed robots.txt decide if
      userAgent can fetch url. Callback function:
      function callback(access, url, reason) { ... }
      where:
      
        access — can this url be fetched. true/false.
        url — target url
        reason — reason for access. Object:
          
            type — valid values: 'statusCode', 'entry', 'defaultEntry', 'noRule'
            entry — an instance of lib/Entry.js:. Only for types: 'entry', 'defaultEntry'
            statusCode — http response status code for url. Only for type 'statusCode'
          
        
      
    function callback(access, url, reason) { ... }access — can this url be fetched. true/false.url — target urlreason — reason for access. Object:
          
            type — valid values: 'statusCode', 'entry', 'defaultEntry', 'noRule'
            entry — an instance of lib/Entry.js:. Only for types: 'entry', 'defaultEntry'
            statusCode — http response status code for url. Only for type 'statusCode'
          
        accesstype — valid values: 'statusCode', 'entry', 'defaultEntry', 'noRule'entry — an instance of lib/Entry.js:. Only for types: 'entry', 'defaultEntry'lib/Entry.js:statusCode — http response status code for url. Only for type 'statusCode'canFetchSync(userAgent, url) — using the parsed robots.txt decide if
      userAgent can fetch url. Return true/false.getCrawlDelay(userAgent) — returns Crawl-delay for the certain userAgentgetSitemaps(sitemaps) — gets Sitemaps from parsed robots.txtgetDisallowedPaths(userAgent) — gets paths explictly disallowed for the user agent specified AND *See LICENSE
    file.Robots.txt Specifications by GoogleRobots.txt parser for pythonA Standard for Robot Exclusion",
  },
  "smartCrawlingUrls": Array [
    "https://npmjs.com/npm/enterprise",
    "https://npmjs.com/features",
    "https://npmjs.com/pricing",
    "https://npmjs.com/support",
    "https://npmjs.com/",
    "https://npmjs.com/logout",
    "https://npmjs.com/~moyuyc",
    "https://npmjs.com/settings/moyuyc/profile",
    "https://npmjs.com/settings/moyuyc/packages",
    "https://npmjs.com/settings/moyuyc/billing",
    "https://npmjs.com/settings/moyuyc/tokens",
    "https://npmjs.com/org/create",
    "https://npmjs.com/search",
    "https://npmjs.com/~ekalinin",
    "https://npmjs.com/policies/security",
    "https://npmjs.com/about",
    "https://npmjs.com/jobs",
    "https://npmjs.com/npm-weekly",
    "https://npmjs.com/policies/terms",
    "https://npmjs.com/policies/conduct",
    "https://npmjs.com/policies/disputes",
    "https://npmjs.com/policies/privacy",
    "https://npmjs.com/policies/receiving-reports",
    "https://npmjs.com/policies/",
  ],
}
`;

exports[`crawl crawl text with js_render 1`] = `
Object {
  "crawlTexts": Object {
    "lvl0": "the-answer",
    "lvl1": null,
    "lvl2": "InstallationUsageReally?License",
    "lvl3": null,
    "lvl4": null,
    "text": "The answer to the question of life, the universe and everything.Yes — I needed a simple module to demonstrate how to import npm packages into Rollup bundles. In particular, this package exposes both a main (UMD format) and module (ES2015 format) in its package.json.mainmoduleMIT",
  },
  "smartCrawlingUrls": Array [],
}
`;

exports[`crawl should crawl nothing 1`] = `
Object {
  "crawlTexts": Object {},
  "smartCrawlingUrls": Array [],
}
`;
